Add ODH-ADR-0005 GitHub Issue Label Standards (#9)
opendatahub-io/architecture-decision-records: ODH-ADR-0005-github-labels-standards.md

| Date           | 2023-04-14 |
| Scope          | |
| Status         | Draft |
| Authors        | [Landon LaSmith](@lavlas) |
| Supersedes     | N/A |
| Superseded by: | N/A |
| Tickets        | |
| Other docs:    | none |

## What

This document will set the the organization standards for the core set of GitHub [Issue Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) that should be supported by every repository in the opendatahub-io organziation.

## Why

[opendatahub-io](https://github.com/opendatahub-io) needs a unified Issue workflow that can support common [queries](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) for all Issue and Pull Request metadata from any repository in the opendatahub-io organization. To support this unified workflow, the required label names must match across all repositories.

## Goals

* Standardize an issue workflow that will be used to show the Issue lifecycle, ownership and category
* Define the label name standard that covers the universal states and metadata that is relevant to the entire opendatahub-io organization

>   * Type: Bug, Feature, Documentation, tracker 

Reviewer: ```suggestion
  * Type: Bug, Feature, Documentation, Tracker 
```

  * Status: To Do, In Progess, In Review, Closed 

>   * Priority 

Reviewer: Needs to be filled in

Author: Since `Priority` is universally known concept for issue tracking, I chose to use the label description table below as the definition

>   * ODH Component: odh-operator, Notebook Controller, Notebooks, Manifests, Data Science Pipelines, Model Mesh, ...

Reviewer: ```suggestion
  * ODH Component: odh-operator, Notebook Controller, Notebooks, Manifests, Data Science Pipelines, Model Serving, AI Explainability, ...
```

>   * ODH Component: odh-operator, Notebook Controller, Notebooks, Manifests, Data Science Pipelines, Model Mesh, ...

Author: This list is reserving common github label names to ensure that they have the same intent across all repositories in the organization.  The intent of these labels are outlined in the table lower in the document

```
This should be used as the indicator for which component should own this issue if located in the centralized `odh-community-issues` repository
```

>   * ODH Component: odh-operator, Notebook Controller, Notebooks, Manifests, Data Science Pipelines, Model Mesh, ...

Reviewer 2: I get that -- but you're about to apply labels to all repos based on this ARD list of items, no? Why do we need ODH Component in all repos... Your ADR even notes that:

> * This ADR will only define the core set of labels that will be supported across all repositories in opendatahub-io.
>  The labels outlined in this document will only be a subset of the available labels in any given repo and will not contain any labels that are isolatedj to an individual repostory workflow

Unless I am wildly misunderstanding this -- what labels you mentioned here are applied to all repos, and they can have more on top of that.

This is not a "core" label -- this is your `odh-community-issues` specific label.

>   * ODH Component: odh-operator, Notebook Controller, Notebooks, Manifests, Data Science Pipelines, Model Mesh, ...

Author: There is still some ambiguity in the lines for the `What` and `Why` sections above
```
This ADR will only define the core set of labels that will be supported across all repositories in opendatahub-io.
```
AND 
```
...the required label names must match across all repositories.
```

These label names are being reserved organization-wide to prevent any conflict in label functionality across repos.  While they will be actively used in `odh-community-issues` they will not be applied to all repositories.

  * SIG: Platform, ML Ops, Developer Experience
  * Extra: Good First Issue

> * Create an `odh-community-issue` repository that will become the centralized Issue triaging location for ODH Component owners to triage new issues and/or transfer to the appropriate component repository

Reviewer 3: Is there a reason to create an additional repository for this purpose? it seems that the `opendatahub-community` repository already contains information about community engagements, SIG, WG, etc. Wouldn't it be better to also use that repository for ticket consolidation and triage as well?

IMHO, having a single repository as the top level community landing point simplifies that workflow and reduces friction for engagement.

Author: That makes sense.  Originally the thought process was that the standalone repo could potentially grow to a bigger automated workflow in the future. As of right now, there is no need for the standlone repo.

## Non-Goals

* This ADR will only define the core set of labels that will be supported across all repositories in opendatahub-io.
  The labels outlined in this document will only be a subset of the available labels in any given repo and will not contain any labels that are isolatedj to an individual repostory workflow
* This is not a mandate that every component or SIG must use the centralized `odh-community-issue` repository to manage the lifecyle of issues relevant to their workflow

## How

Across all repositories in the opendatahub-io organization, we will create the labels below with the expectation that the workflows they outline will be universal across all opendatahub-io repositories

> | Label                         | Description     |

Reviewer 4: What about `kind/security` for reporting security vulnerabilities?

| ----------------------------- | ---------------- |
| `tracker`             | Non-completable ticket; used for tracking work - akin to a Jira Epic |
| `untriaged`           | Indicates the newly create issue has not been triaged yet |
| `kind/bug`            | Indicates an unexpected problem or unintended behavior |
| `kind/enhancement`    | New functionality request (existing augments or new additions) |
| `kind/documentation`  | Improvements or additions to documentation |

> | `NEEDINFO`            | Further information is requested to unblock any progress on the issue |

Reviewer 2: ```suggestion
| `needs-info`            | Further information is requested to unblock any progress on the issue |
```
Any reason this is not just snake-case like the others? (I guess just `good-first-issue`)

Author: This is an older standard from bugzilla `NEEDINFO` in all caps.  There is no reason it needs to be all caps

| `priority/normal`     | An issue with the product; fix when possible                |
| `priority/blocker`    | Critical issue that needs to be fixed asap; blocks up coming releases                |
| `priority/low`        | An issue with the product that doesn't impact the user much or not at all (ie tech debt)    |
| `priority/high`       | Important issue that needs to be resolved asap. Releases should not have too many of these. |
| `good-first-issue`    | Good for newcomers |
| `odh-component/*`     | Name of the odh-component that owns this issue.  This should be used as the indicator for which component should own this issue if located in the centralized `odh-community-issues` repository |
| `sig/*`               | Name of the Special Interest Group in charge of this subject matter|

> | `wg/*`                | Name of the Working Group that should be assigned to this issue |

Author: SIGs and WG are organization-wide entities and would affect all repositories in the org. This list is reserving common github label names to ensure that they have the same intent across all repositories in the organization.

> | `wg/*`                | Name of the Working Group that should be assigned to this issue |

Reviewer 2: Fair -- I suppose I don't see the Dashboard ever using either sig or wg -- but I see no harm in collectively aligning.

I still disagree with `odh-component` being in every repo though... that still is a confusing thing for me to grok.

Additional labels maybe reserved depending on certain bots or apps running in the organization.

It is assumed that all new issues will have the `untriaged` label until it is reviewed.  Once an issue is triaged, a `kind/*` and `priority/*` label should be added and the issue will follow the issue worklow outlined in the component repo where it is located.

## Open Questions

> * Does the workflow for the `odh-community-issues` repository need to be defined here OR in the repo README when it is created

Reviewer: How complex do we think the workflow will be? If we're just discussing high level plans I think it's worth including in here, perhaps even a basic flowchart.

Reviewer 2: I imagine we should definitely have a very light process here... this is for external users to join in... entry should be as frictionless as possible.

A couple issue templates? Bug, Feature, maybe(?) Documentation... No PRs.

I wonder how we will properly engage everyone -- maybe this is just a process for us to triage like the Jira Dashbaord or any upstream processes that we would use. Could have a little friction on our end to triage, but it shouldn't be hard for either side (new issue creator + triager) to collaborate.
